### PR TITLE
[Proposal] SOAR-0004: Streaming request and response bodies

### PR DESCRIPTION
**Pull request by czechboy0**
### Motivation

See the proposal at https://github.com/czechboy0/swift-openapi-generator/blob/hd-soar-0004/Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md.

### Modifications

N/A

### Result

N/A

### Test Plan

N/A

### CODE REVIEW

**Comment by gjcairo** [comment]
This proposal is now In Review, which will run until 15th September - feel free to leave your feedback either on this PR or on the corresponding [Swift Forums post](https://forums.swift.org/t/proposal-soar-0004-streaming-request-and-response-bodies/67184).

Thanks, and thanks @czechboy0 for the write-up!

**Comment by gjcairo** [comment]
Overall really like the shape of things here, great job Honza 🥳 I've only got two pretty minor comments:

1. As a general rule, we tend to favour extending `Bar` with `init(foo: Foo)` when `Foo` can be converted into a `Bar`, instead of having `Foo.asBar()` methods. I wonder if we should make the following changes - have you thought about this?
 - `HTTPBody.collectAsString(upTo maxBytes: Int) async throws -> String` -> `String.init(collecting httpBody: HTTPBody, upTo maxBytes: Int) async throws`
 - `HTTPBody.collectAsData(upTo maxBytes: Int) async throws -> Data` -> `Data.init(collecting httpBody: HTTPBody, upTo maxBytes: Int) async throws`
 - `StringProtocol`'s extension to add `asBodyChunk` -> `HTTPBody.ByteChunk.init(_ string: String)`
2. This one really is about the proposal itself and quite minor, but if `collect(upTo: .max)` is discouraged (as stated in the text), should we use it in examples in the proposal? Maybe we should steer clear of what we deem bad practices in sample code.

**Comment by czechboy0** [comment]
> Overall really like the shape of things here, great job Honza 🥳 I've only got two pretty minor comments:
> 
> 1. As a general rule, we tend to favour extending `Bar` with `init(foo: Foo)` when `Foo` can be converted into a `Bar`, instead of having `Foo.asBar()` methods. I wonder if we should make the following changes - have you thought about this?
> 
> * `HTTPBody.collectAsString(upTo maxBytes: Int) async throws -> String` -> `String.init(collecting httpBody: HTTPBody, upTo maxBytes: Int) async throws`
> * `HTTPBody.collectAsData(upTo maxBytes: Int) async throws -> Data` -> `Data.init(collecting httpBody: HTTPBody, upTo maxBytes: Int) async throws`

I quite like that. I guess it's less common for async initializers, but that shouldn't stop us. While it's a little less discoverable, I think the consistency with general Swift guidelines is more important here. I'll make that change.

> * `StringProtocol`'s extension to add `asBodyChunk` -> `HTTPBody.ByteChunk.init(_ string: String)`

This is an internal, not public, method, it just showed up there as it's inlinable. I'll see later about changing it, but probably not before the review ends.

> 2. This one really is about the proposal itself and quite minor, but if `collect(upTo: .max)` is discouraged (as stated in the text), should we use it in examples in the proposal? Maybe we should steer clear of what we deem bad practices in sample code.

Ha - keeping me honest. Yes, I'll update it 🙂 

Thanks, @gjcairo! 🙏

**Comment by czechboy0** [comment]
@gjcairo I applied all your feedback, please take a look at v1.3: https://github.com/czechboy0/swift-openapi-generator/blob/hd-soar-0004/Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md

File: Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md [hunk 1]
```diff
@@ -305,24 +308,24 @@ A reminder that the exact spelling of integrating `HTTPBody` into the transport
...
 /// If you need to collect the whole body before processing it, use one of
-/// the convenience `collect` methods on `HTTPBody`.
+/// the convenience initializers on the target types that take an `HTTPBody`.
 ///
-/// To get all the bytes, use:
+/// To get all the bytes, use the initializer on `ArraySlice<UInt8>` or `[UInt8]`:
 ///
 /// ```swift
-/// let buffer = try await body.collect(upTo: 2 * 1024 * 1024)
+/// let buffer = try await ArraySlice(collecting: body, upTo: 2 * 1024 * 1024)
```
Comment (gjcairo) [inline Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md:316]: Tiny nit: you use this new initialiser here but I don't think it's included in the code snippet below.

File: Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md [hunk 2]
```diff
@@ -470,71 +505,52 @@ extension HTTPBody {
...
-    /// the provided maximum number of bytes, converts it to string from
-    /// the UTF-8 bytes, and returns it.
+    /// Creates a string by accumulating the full body in-memory into a single buffer up to
+    /// the provided maximum number of bytes, converting it to string using the provided encoding.
     /// - Parameters:
+    ///   - body: The HTTP body to collect.
     ///   - maxBytes: The maximum number of bytes this method is allowed
     ///     to accumulate in memory before it throws an error.
-    /// - Throws: `TooManyBytesError` if the the body contains more
+    /// - Throws: `TooManyBytesError` if the the sequence contains more
```
Comment (gjcairo) [inline Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md:551]: Nit: extra "the"
```suggestion
    /// - Throws: `TooManyBytesError` if the sequence contains more
```
Also, isn't `body` more accurate here?

File: Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md (hunk 1 above)
Comment (czechboy0) [reply Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md:316]: Was already fixed.

File: Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md (hunk 2 above)
Comment (czechboy0) [reply Sources/swift-openapi-generator/Documentation.docc/Proposals/SOAR-0004.md:551]: Fixed.

**Comment by gjcairo** [comment]
The review period is now over, and the proposal has been accepted. SOAR-0004 is now **Ready for Implementation**.

**Comment by czechboy0** [comment]
Landing the proposals to avoid more conflicts, will update their status once the implementation lands.

**Pull request merged**